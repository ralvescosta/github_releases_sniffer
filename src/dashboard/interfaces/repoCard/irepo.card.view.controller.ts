import React from 'react';

export interface IRepoCardViewController {
  modalControl: boolean;
  setModalControl: React.Dispatch<React.SetStateAction<boolean>>;
  openModal: (repositoryId: number) => void;
}
