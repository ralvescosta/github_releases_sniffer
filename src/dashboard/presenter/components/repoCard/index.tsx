import React from 'react';
import {View, Text, Alert} from 'react-native';
import {TouchableOpacity} from 'react-native-gesture-handler';
import FontAwesome from 'react-native-vector-icons/FontAwesome';

import {styles} from './styles';

type Props = {
  repository: any;
};

export const RepoCard = ({repository}: Props) => {
  function pressedCard() {
    Alert.alert('Github Sniffer', 'Do you really want removing this repository sniffed?', [
      {
        text: 'cancel',
        style: 'cancel',
      },
      {
        text: 'OK',
        onPress: async () => {},
      },
    ]);
  }

  return (
    <TouchableOpacity style={styles.repoCard} onPress={pressedCard}>
      <View style={styles.cardLeft}>
        <View style={styles.leftHeader}>
          <Text numberOfLines={2} style={styles.headerName}>
            {repository.fullName}
          </Text>
          <Text numberOfLines={3} ellipsizeMode="tail" style={styles.headerDescription}>
            {repository.description}
          </Text>
        </View>

        <View style={styles.leftContent}>
          <View style={styles.contentDetails}>
            <FontAwesome name="star" color="#777" size={15} />
            <Text numberOfLines={1} style={styles.detailsText}>
              {repository.stargazersCount}
            </Text>
          </View>

          <View style={styles.contentDetails}>
            <FontAwesome name="code-fork" color="#777" size={15} />
            <Text numberOfLines={1} style={styles.detailsText}>
              {repository.forks}
            </Text>
          </View>

          <View style={styles.contentDetails}>
            <FontAwesome name="bug" color="#777" size={15} />
            <Text numberOfLines={1} style={styles.detailsText}>
              {repository.openIssues}
            </Text>
          </View>
        </View>
      </View>

      <View style={styles.cardRight}>
        <View style={styles.rightRelease}>
          <Text style={styles.releaseTitle}>Last Release: </Text>
          <Text style={styles.releaseTag}>{repository.lastRelease}</Text>
        </View>
        <FontAwesome name="bell" color="#d4d4d4" size={30} />
      </View>
    </TouchableOpacity>
  );
};
