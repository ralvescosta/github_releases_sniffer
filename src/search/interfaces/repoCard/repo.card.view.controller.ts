import {useState, useContext} from 'react';
import {IRepoCardViewController} from './irepo.card.view.controller';

import {GlobalContext} from '../../../core/context/sniffed.repositories.context';

import {ResultSearchGithubRepositoryEntity} from '../../bussiness/entities/result.search.github.repository.entity';
import {ISaveRepositoryToSnifferUsecase} from '../../bussiness/usecases/isave.repositrory.to.sniffer.usecase';
import {IRemoveRepositoryCheckedAsSnifferUsecase} from '../../bussiness/usecases/iremove.repository.checked.as.sniffer.usecase';

export class RepoCardViewController implements IRepoCardViewController {
  public repository: ResultSearchGithubRepositoryEntity;

  public switchState: boolean;
  public setSwitchState: React.Dispatch<React.SetStateAction<boolean>>;

  public loading: boolean;
  public setLoading: React.Dispatch<React.SetStateAction<boolean>>;

  public context = useContext(GlobalContext);

  constructor(
    repository: ResultSearchGithubRepositoryEntity,
    private readonly _saveToSnifferUsecase: ISaveRepositoryToSnifferUsecase,
    private readonly _removeSniffedUsecase: IRemoveRepositoryCheckedAsSnifferUsecase,
  ) {
    this.repository = repository;
    [this.switchState, this.setSwitchState] = useState<boolean>(this.repository.checked);
    [this.loading, this.setLoading] = useState<boolean>(false);
  }

  public async saveRepositoryToSniffer(): Promise<void> {
    this.setLoading(true);
    try {
      const sniffed = await this._saveToSnifferUsecase.saveToSniffer(this.repository);
      this.context.setSniffedRepositories([...this.context.sniffedRepositories, sniffed]);
      this.setSwitchState(true);
    } catch (err) {}

    this.setLoading(false);
  }

  public async removeSnifferRepository() {
    this.setLoading(true);

    const result = await this._removeSniffedUsecase.remove(this.repository.id, this.context.sniffedRepositories);
    this.context.setSniffedRepositories(result);

    this.setSwitchState(false);
    this.setLoading(false);
  }
}
