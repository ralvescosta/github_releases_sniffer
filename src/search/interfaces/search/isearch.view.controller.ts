import React from 'react';
import {ResultSearchGithubRepositoryEntity} from '../../bussiness/entities/result.search.github.repository.entity';

export interface ISearchViewController {
  repositoryName: React.MutableRefObject<string>;

  isLoading: boolean;
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;

  foundRepositories: React.MutableRefObject<ResultSearchGithubRepositoryEntity[]>;

  context: any;

  searchRepository: () => Promise<void>;
}
