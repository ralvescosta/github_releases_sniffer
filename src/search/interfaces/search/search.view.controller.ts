import {useState, useRef, useContext} from 'react';
import {Keyboard} from 'react-native';

import {GlobalContext} from '../../../core/context/sniffed.repositories.context';
import {ISearchViewController} from './isearch.view.controller';

import {ISearchGithubRepositoryUsecase} from '../../bussiness/usecases/isearch.github.repository.usecase';
import {ResultSearchGithubRepositoryEntity} from '../../bussiness/entities/result.search.github.repository.entity';

export class SearchViewController implements ISearchViewController {
  public repositoryName: React.MutableRefObject<string>;

  public isLoading: boolean;
  public setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;

  public foundRepositories: React.MutableRefObject<ResultSearchGithubRepositoryEntity[]>;

  public context = useContext(GlobalContext);
  public isNetworkAvailable = this.context.isNetworkAvailable;

  constructor(private readonly _searchRepositoryUsecase: ISearchGithubRepositoryUsecase) {
    this.repositoryName = useRef('');
    [this.isLoading, this.setIsLoading] = useState<boolean>(false);

    this.foundRepositories = useRef<ResultSearchGithubRepositoryEntity[]>([]);
  }

  public async searchRepository(): Promise<void> {
    if (!this.repositoryName.current) {
      throw new Error();
    }

    if (!this.isNetworkAvailable) {
      return;
    }

    this.setIsLoading(true);
    Keyboard.dismiss();

    try {
      const result = await this._searchRepositoryUsecase.search(this.repositoryName.current, this.context.sniffedRepositories);
      this.foundRepositories.current = result;
    } catch (err) {}

    this.setIsLoading(false);
  }
}
