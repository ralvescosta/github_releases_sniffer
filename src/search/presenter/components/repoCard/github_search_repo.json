{
  "total_count": 1,
  "incomplete_results": false,
  "items": [
    {
      "id": 29028775,
      "node_id": "MDEwOlJlcG9zaXRvcnkyOTAyODc3NQ==",
      "name": "react-native",
      "full_name": "facebook/react-native",
      "private": false,
      "owner": {
        "login": "facebook",
        "id": 69631,
        "node_id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "avatar_url": "https://avatars3.githubusercontent.com/u/69631?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/facebook",
        "html_url": "https://github.com/facebook",
        "followers_url": "https://api.github.com/users/facebook/followers",
        "following_url": "https://api.github.com/users/facebook/following{/other_user}",
        "gists_url": "https://api.github.com/users/facebook/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/facebook/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/facebook/subscriptions",
        "organizations_url": "https://api.github.com/users/facebook/orgs",
        "repos_url": "https://api.github.com/users/facebook/repos",
        "events_url": "https://api.github.com/users/facebook/events{/privacy}",
        "received_events_url": "https://api.github.com/users/facebook/received_events",
        "type": "Organization",
        "site_admin": false
      },
      "html_url": "https://github.com/facebook/react-native",
      "description": "A framework for building native apps with React.",
      "fork": false,
      "url": "https://api.github.com/repos/facebook/react-native",
      "forks_url": "https://api.github.com/repos/facebook/react-native/forks",
      "keys_url": "https://api.github.com/repos/facebook/react-native/keys{/key_id}",
      "collaborators_url": "https://api.github.com/repos/facebook/react-native/collaborators{/collaborator}",
      "teams_url": "https://api.github.com/repos/facebook/react-native/teams",
      "hooks_url": "https://api.github.com/repos/facebook/react-native/hooks",
      "issue_events_url": "https://api.github.com/repos/facebook/react-native/issues/events{/number}",
      "events_url": "https://api.github.com/repos/facebook/react-native/events",
      "assignees_url": "https://api.github.com/repos/facebook/react-native/assignees{/user}",
      "branches_url": "https://api.github.com/repos/facebook/react-native/branches{/branch}",
      "tags_url": "https://api.github.com/repos/facebook/react-native/tags",
      "blobs_url": "https://api.github.com/repos/facebook/react-native/git/blobs{/sha}",
      "git_tags_url": "https://api.github.com/repos/facebook/react-native/git/tags{/sha}",
      "git_refs_url": "https://api.github.com/repos/facebook/react-native/git/refs{/sha}",
      "trees_url": "https://api.github.com/repos/facebook/react-native/git/trees{/sha}",
      "statuses_url": "https://api.github.com/repos/facebook/react-native/statuses/{sha}",
      "languages_url": "https://api.github.com/repos/facebook/react-native/languages",
      "stargazers_url": "https://api.github.com/repos/facebook/react-native/stargazers",
      "contributors_url": "https://api.github.com/repos/facebook/react-native/contributors",
      "subscribers_url": "https://api.github.com/repos/facebook/react-native/subscribers",
      "subscription_url": "https://api.github.com/repos/facebook/react-native/subscription",
      "commits_url": "https://api.github.com/repos/facebook/react-native/commits{/sha}",
      "git_commits_url": "https://api.github.com/repos/facebook/react-native/git/commits{/sha}",
      "comments_url": "https://api.github.com/repos/facebook/react-native/comments{/number}",
      "issue_comment_url": "https://api.github.com/repos/facebook/react-native/issues/comments{/number}",
      "contents_url": "https://api.github.com/repos/facebook/react-native/contents/{+path}",
      "compare_url": "https://api.github.com/repos/facebook/react-native/compare/{base}...{head}",
      "merges_url": "https://api.github.com/repos/facebook/react-native/merges",
      "archive_url": "https://api.github.com/repos/facebook/react-native/{archive_format}{/ref}",
      "downloads_url": "https://api.github.com/repos/facebook/react-native/downloads",
      "issues_url": "https://api.github.com/repos/facebook/react-native/issues{/number}",
      "pulls_url": "https://api.github.com/repos/facebook/react-native/pulls{/number}",
      "milestones_url": "https://api.github.com/repos/facebook/react-native/milestones{/number}",
      "notifications_url": "https://api.github.com/repos/facebook/react-native/notifications{?since,all,participating}",
      "labels_url": "https://api.github.com/repos/facebook/react-native/labels{/name}",
      "releases_url": "https://api.github.com/repos/facebook/react-native/releases{/id}",
      "deployments_url": "https://api.github.com/repos/facebook/react-native/deployments",
      "created_at": "2015-01-09T18:10:16Z",
      "updated_at": "2020-09-07T09:33:12Z",
      "pushed_at": "2020-09-07T08:11:37Z",
      "git_url": "git://github.com/facebook/react-native.git",
      "ssh_url": "git@github.com:facebook/react-native.git",
      "clone_url": "https://github.com/facebook/react-native.git",
      "svn_url": "https://github.com/facebook/react-native",
      "homepage": "https://reactnative.dev",
      "size": 551737,
      "stargazers_count": 89929,
      "watchers_count": 89929,
      "language": "JavaScript",
      "has_issues": true,
      "has_projects": true,
      "has_downloads": true,
      "has_wiki": true,
      "has_pages": true,
      "forks_count": 19905,
      "mirror_url": null,
      "archived": false,
      "disabled": false,
      "open_issues_count": 1020,
      "license": {
        "key": "other",
        "name": "Other",
        "spdx_id": "NOASSERTION",
        "url": null,
        "node_id": "MDc6TGljZW5zZTA="
      },
      "forks": 19905,
      "open_issues": 1020,
      "watchers": 89929,
      "default_branch": "master",
      "score": 1.0
    }
  ]
}