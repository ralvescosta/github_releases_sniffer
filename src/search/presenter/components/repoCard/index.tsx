import React, {useState} from 'react';
import {View, Text, Switch} from 'react-native';

import {styles} from './styles';
import {ResultSearchGithubRepositoryEntity} from '../../../bussiness/entities/result.search.github.repository.entity';

type Props = {
  repository: ResultSearchGithubRepositoryEntity;
};

export const RepoCard: React.FC<Props> = ({repository}) => {
  const [switchState, setSwitchState] = useState(false);
  return (
    <View style={styles.repoCard}>
      <View style={styles.cardLeft}>
        <View style={styles.leftHeader}>
          <Text numberOfLines={2} style={styles.headerName}>
            {repository.fullName}
          </Text>
          <Text numberOfLines={2} ellipsizeMode="tail" style={styles.headerDescription}>
            {repository.description}
          </Text>
        </View>
      </View>

      <View style={styles.cardRight}>
        <Switch value={switchState} onValueChange={setSwitchState} />
      </View>
    </View>
  );
};
