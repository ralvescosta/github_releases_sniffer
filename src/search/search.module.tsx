import React from 'react';
import {SearchScreen} from './presenter/search.screen';

import {SearchGithubRepoRepository} from './infrastructure/repositories/search.github.repo.repository';
import {SearchGithubRepositoryUsecase} from './application/usecases/search.github.repository.usecase';
import {SearchViewController} from './interfaces/search.view.controller';

import {RepoCard} from './presenter/components/repoCard';
import {RepoCardViewController} from './interfaces/repo.card.view.controller';

import {SaveRepositoryToSnifferUsecase} from './application/usecases/save.repositrory.to.sniffer.usecase';
import {SaveLocallySniffedRepository} from './infrastructure/repositories/save.locally.sniffed.repository';
import {GetLastSniffedReleaseRepository} from './infrastructure/repositories/get.last.sniffed.release.repository';

import {RemoveRepositoryCheckedAsSnifferUsecase} from './application/usecases/remove.repository.checked.as.sniffer.usecase';
import {UpdateLocallySniffedRepository} from './infrastructure/repositories/update.locally.sniffed.repository';

const RepositoryCardComponent = ({repository}: any) => {
  const viewController = new RepoCardViewController(
    repository,
    SaveRepositoryToSnifferUsecase.getInstance(SaveLocallySniffedRepository.getInstance(), GetLastSniffedReleaseRepository.getInstance()),
    RemoveRepositoryCheckedAsSnifferUsecase.getInstance(UpdateLocallySniffedRepository.getInstance()),
  );

  return <RepoCard viewController={viewController} />;
};

export const SearchScreenModule = () => {
  const searchGithubRepoRepository = new SearchGithubRepoRepository();
  const searchGithubRepositoryUsecase = new SearchGithubRepositoryUsecase(searchGithubRepoRepository);

  const searchViewController = new SearchViewController(searchGithubRepositoryUsecase);

  return <SearchScreen viewController={searchViewController} RepoCard={RepositoryCardComponent} />;
};
