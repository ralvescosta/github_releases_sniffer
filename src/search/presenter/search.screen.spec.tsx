import React from 'react';
import {render} from '@testing-library/react-native';

import {SearchScreen} from './search.screen';
import {SearchViewControllerSpy} from './__test__/search.view.controller.spy';
import {RepoCardViewControllerSpy} from './__test__/repo.card.view.controller.spy';
import {RepoCard} from './components/repoCard';

function makeSut() {
  const Wrapper = () => {
    const searchController = new SearchViewControllerSpy();
    const repoController = new RepoCardViewControllerSpy();

    const RepoWrapper = () => {
      return <RepoCard viewController={repoController} />;
    };
    return <SearchScreen viewController={searchController} RepoCard={RepoWrapper} />;
  };

  return render(<Wrapper />);
}

describe('Search Screen', () => {
  it('Should return correctly', () => {
    const sut = makeSut();

    expect(sut.toJSON()).toMatchSnapshot();
  });
});
