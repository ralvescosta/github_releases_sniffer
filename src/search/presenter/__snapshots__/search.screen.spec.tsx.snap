// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Screen Should return correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f4f4f4",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#24292e",
        "height": 50,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 20,
          "elevation": 5,
          "flexDirection": "row",
          "maxWidth": "90%",
          "paddingHorizontal": 7,
          "position": "absolute",
          "top": 15,
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Repositorio"
        rejectResponderTermination={true}
        style={
          Object {
            "width": "85%",
          }
        }
        underlineColorAndroid="transparent"
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "opacity": 1,
            "padding": 15,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 15,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï€‚
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "marginTop": 15,
        "paddingBottom": 45,
      }
    }
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View />
  </RCTScrollView>
</View>
`;
